Patched hypre to better add flags based on compiler.

frobnitzem opened the pull request:
This new patch is required to build hypre with openmp.
The root cause is explained here: https://github.com/hypre-space/hypre/pull/225
hypre's build system doesn't add the -fopenmp flag unless the compiler name matches some preset ones, and absolute paths break that logic.

There have been a few changes to the configure script, so the line numbers might be off.  It might be better to rephrase this patch with a sed command.  I don't know how to do that.

frobnitzem committed:
Patched hypre to better add flags based on compiler.

adamjstewart reviewed (approved):
@ulrikeyang @osborn9 @balay can you review?

osborn9 reviewed (changes requested):
Thanks for this fix! Could you please add a `when` condition to line 72 of package.py (i.e. patch('detect-compiler.patch', when='@2.15.0:')) like [package_diff.txt](https://github.com/spack/spack/files/5440862/package_diff.txt), since line numbers are different for old versions (even though your etc. fix won't apply for the old versions, at least the patch won't fail).

frobnitzem committed:
Update package.py

This file seems to have lots of edits, so the patch may succeed with offsets.  Has anyone checked with spack patch to be sure it'll work with versions 2.15 - 2.20?
